cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,384 @@
+
+        Raises:
+            NotImplementedError: The new circuit contains measurements, but
+                this operation has repetitions.
+        """
+        if protocols.is_measurement(new_circuit) and self._repetitions != 1:
+            raise NotImplementedError('Measurements cannot be added to looped circuits.')
+        new_op = self.copy()
+        new_op._circuit = new_circuit.freeze()
+        return new_op
+
+    def repeat(
+        self,
+        repetitions: int,
+        modulus: int = 0,
+        allow_invert: bool = False,
+        validate_modulus: bool = False,
+    ) -> 'CircuitOperation':
+        """Returns a copy of this operation repeated 'repetitions' times.
+
+        Args:
+            repetitions: Number of times this operation should repeat. This
+                is multiplied with any pre-existing repetitions.
+            modulus: If this is specified, the final repetition count will be
+                mapped to the range [0, modulus), or (-modulus/2, modulus/2]
+                if `allow_invert` is True.
+            allow_invert: For use with `modulus`. If this is True, repetitions
+                will map to the range (-modulus/2, modulus/2]. This is useful
+                for minimizing the total number of operations in a circuit.
+            validate_modulus: For use with `modulus`. If this is True, the
+                function will raise an error if the provided modulus does not
+                match the actual modulus of this operation.
+
+        Returns:
+            A copy of this operation repeated 'repetitions' times.
+
+        Raises:
+            TypeError: `repetitions` is not an integer value.
+            NotImplementedError: The operation contains measurements and
+                cannot have repetitions.
+            ValueError: `validate_modulus` was set to True, but raising this
+                operation to `modulus` does not produce the identity.
+        """
+        if not isinstance(repetitions, int):
+            raise TypeError('Only integer repetitions are allowed.')
+        if protocols.is_measurement(self.circuit):
+            raise NotImplementedError('Loops over measurements are not supported.')
+        new_op = self.copy()
+        new_op._repetitions *= repetitions
+        if modulus:
+            if validate_modulus and not linalg.allclose_up_to_global_phase(
+                protocols.unitary(self.circuit * modulus), np.eye(np.prod(self.circuit.qid_shape()))
+            ):
+                raise ValueError('Raising the circuit to "modulus" must produce the identity.')
+            new_op._repetitions %= modulus
+            if allow_invert and new_op._repetitions > modulus // 2:
+                new_op._repetitions -= modulus
+        return new_op
+
+    def __pow__(self, power: int) -> 'CircuitOperation':
+        return self.repeat(power)
+
+    def with_qubit_mapping(self, qubit_map: Dict['cirq.Qid', 'cirq.Qid']) -> 'CircuitOperation':
```

Comment:
`Circuit` has a similar method `transform_qubits` which takes a `Callable[[Qid], Qid]` instead of a dict. I think it'd be nice to harmonize these APIs. Ideally `transform_qubits` would accept `Mapping[Qid, Qid]` in addition to callable, since many callsite wrap a lambda around a dict anyway. Could also accept both callable and dict here. It'd be great if we could name the functions similarly as well; I like `with` since it produces a new instance, so maybe we could add a `with_qubit_mapping` on circuit too?

Reply:
This sounds reasonable to me, but `Operation` also has a `transform_qubits` method. Let's open a separate PR for this rather than addressing it here. For now, I'll update `with_qubit_mapping` to accept functions.

One other thought on this: `with_qubit_mapping` does not directly modify the circuit, unlike `transform_qubits`. I think having separate names to differentiate between these behaviors is worthwhile.

Reply:
Opened #3611 to track the `Circuit` and `Operation` change.